src/dbnode/storage/repair.go

```diff
@@ -441,3 +480,194 @@ func (r repairerNoOp) Start()        {}
 func (r repairerNoOp) Stop()         {}
 func (r repairerNoOp) Repair() error { return nil }
 func (r repairerNoOp) Report()       {}
+
+func (r shardRepairer) shadowCompare(
+	start time.Time,
+	end time.Time,
+	localMetadataBlocks block.FetchBlocksMetadataResults,
+	session client.AdminSession,
+	shard databaseShard,
+	nsCtx namespace.Context,
+) error {
+	dice, err := dice.NewDice(r.rpopts.DebugShadowComparisonsPercentage())
+	if err != nil {
+		return fmt.Errorf("err creating shadow comparison dice: %v", err)
+	}
+
+	var localM, peerM *dynamic.Message
+	if nsCtx.Schema != nil {
+		// Only required if a schema (proto feature) is present. Reset between uses.
+		localM = dynamic.NewMessage(nsCtx.Schema.Get().MessageDescriptor)
+		peerM = dynamic.NewMessage(nsCtx.Schema.Get().MessageDescriptor)
+	}
+
+	tmpCtx := context.NewContext()
+	compareResultFunc := func(result block.FetchBlocksMetadataResult) error {
+		seriesID := result.ID
+		peerSeriesIter, err := session.Fetch(nsCtx.ID, seriesID, start, end)
+		if err != nil {
+			return err
+		}
+		defer peerSeriesIter.Close()
+
+		tmpCtx.Reset()
+		defer tmpCtx.BlockingClose()
+		localSeriesDataBlocks, err := shard.ReadEncoded(tmpCtx, seriesID, start, end, nsCtx)
+		if err != nil {
+			return err
+		}
+		localSeriesSliceOfSlices := xio.NewReaderSliceOfSlicesFromBlockReadersIterator(localSeriesDataBlocks)
+		localSeriesIter := r.opts.MultiReaderIteratorPool().Get()
+		localSeriesIter.ResetSliceOfSlices(localSeriesSliceOfSlices, nsCtx.Schema)
+
+		var (
+			i             = 0
+			foundMismatch = false
+		)
+		for localSeriesIter.Next() {
+			if !peerSeriesIter.Next() {
+				r.logger.Error(
+					"series had next locally, but not from peers",
+					zap.String("namespace", nsCtx.ID.String()),
+					zap.Time("start", start),
+					zap.Time("end", end),
+					zap.String("series", seriesID.String()),
+					zap.Error(peerSeriesIter.Err()),
+				)
+				foundMismatch = true
+				break
+			}
+
+			localDP, localUnit, localAnnotation := localSeriesIter.Current()
+			peerDP, peerUnit, peerAnnotation := peerSeriesIter.Current()
+
+			if !localDP.Timestamp.Equal(peerDP.Timestamp) {
```

Comment:
Might be cleaner to make an equality function that wraps these equality checks together. I could see this logic being reused elsewhere too.

Reply:
lol just realized they already have one

Reply:
Heh nice. Would it be possible to have an equality func for the `(ts.Datapoint, xtime.Unit, ts.Annotation)` tuple too? That might be useful since `encoding.Iterator` usage is pretty common.

Reply:
Would be pretty rare to need that I think (this is pretty much the only case) and it would also make the default behavior of the function slow since you'd have to compare two (potentially) large byte slices. Would prefer to leave that out for now unless we have a concrete need for it (also we have better logs this way otherwise I'd have to log the annotation every time)